http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java

```diff
@@ -0,0 +1,370 @@
+import software.amazon.awssdk.http.crt.internal.AwsCrtAsyncHttpStreamAdapter;
+import software.amazon.awssdk.utils.AttributeMap;
+import software.amazon.awssdk.utils.IoUtils;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.http.SdkHttpUtils;
+
+/**
+ * An implementation of {@link SdkHttpClient} that uses the AWS Common Runtime (CRT) Http Client to communicate with
+ * Http Web Services. This client is asynchronous and uses non-blocking IO.
+ *
+ * <p>This can be created via {@link #builder()}</p>
+ */
+@SdkPublicApi
+public class AwsCrtAsyncHttpClient implements SdkAsyncHttpClient {
+    private static final Logger log = Logger.loggerFor(AwsCrtAsyncHttpClient.class);
+    private static final String HOST_HEADER = "Host";
+    private static final String CONTENT_LENGTH = "Content-Length";
+    private static final String CONNECTION = "Connection";
+    private static final String KEEP_ALIVE = "keep-alive";
+    private static final String AWS_COMMON_RUNTIME = "AwsCommonRuntime";
+    private static final int DEFAULT_STREAM_WINDOW_SIZE = 16 * 1024 * 1024; // 16 MB
+
+    private final Map<URI, HttpClientConnectionManager> connectionPools = new ConcurrentHashMap<>();
+    private final LinkedList<CrtResource> ownedSubResources = new LinkedList<>();
+    private final AtomicBoolean isClosed = new AtomicBoolean(false);
+    private final ClientBootstrap bootstrap;
+    private final SocketOptions socketOptions;
+    private final TlsContextOptions tlsContextOptions;
+    private final TlsContext tlsContext;
+    private final int windowSize;
+    private final int maxConnectionsPerEndpoint;
+
+
+    public AwsCrtAsyncHttpClient(DefaultBuilder builder, AttributeMap config) throws CrtRuntimeException {
+        int maxConns = config.get(SdkHttpConfigurationOption.MAX_CONNECTIONS);
+
+        Validate.isPositive(maxConns, "maxConns");
+        Validate.isPositive(builder.eventLoopSize, "eventLoopSize");
+        Validate.notNull(builder.cipherPreference, "cipherPreference");
+        Validate.isPositive(builder.windowSize, "windowSize");
+
+        /**
+         * Must call own() in same order that CrtResources are created in, so that they will be closed in reverse order.
+         *
+         * Do NOT use Dependency Injection for Native CrtResources. It's possible to crash the JVM Process if Native
+         * Resources are closed in the wrong order (Eg closing the Bootstrap/Threadpool when there are still open
+         * connections). By creating and owning our own Native CrtResources we can guarantee that things are shutdown
+         * in the correct order.
+         */
+
+        try (EventLoopGroup elg = new EventLoopGroup(builder.eventLoopSize);
+            HostResolver hr = new HostResolver(elg)) {
+
+            bootstrap = own(new ClientBootstrap(elg, hr));
+            socketOptions = own(new SocketOptions());
```

Comment:
why are the `elg` and `hr` "try-with-resources", while the rest of the CrtResources are "owned"?

Reply:
Shouldn't have to worry about this now that we aren't creating the EventLoopGroup or HostResolver inside of this constructor.  Previously, it was because I was brining over a chunk of code from a now removed ClientBootStrap constructor that created these objects internally.  Because of that, and because they were being passed to the ClientBootstrap, it didn't seem necessary to track them with "own".  (But I might have been wrong!)